### PR TITLE
Add App launcher shortcut for contact journal (EXPOSUREAPP-4659)

### PR DESCRIPTION
**Pull request by jurajkusnier**
## Introduce App shortcut
App shortcut should provide direct access to the contact journal (current day)

The shortcut menu should not be visible if
- onboarding was not finished yet
- submission process was started (from entering "Your test result is available" until returning back to the home screen)

## Testing
- New test menu item "Onboarding" was introduced to simplify simulating edge cases
- Delta onboarding/ New features/ Contact diary onboarding should be displayed before entering the contact journal

### CODE REVIEW

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+@Singleton
+class AppShortcutsHelper @Inject constructor(
+    private val shortcutManager: ShortcutManager,
+    @AppContext private val context: Context
+) {
+
+    fun createAppShortcut() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+            val shortcut = ShortcutInfo.Builder(context, CONTACT_DIARY_SHORTCUT_ID)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt:24]: Please use `ShortcutInfoCompat` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                .setShortLabel(context.getString(R.string.app_shortcut_contact_diary_title))
+                .setLongLabel(context.getString(R.string.app_shortcut_contact_diary_title))
+                .setIcon(Icon.createWithResource(context, R.drawable.ic_contact_diary_shortcut_icon))
+                .setIntent(createContactDiaryIntent())
+                .build()
+
+            shortcutManager.dynamicShortcuts = listOf(shortcut)
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt:31]: ShortcutManagerCompat.addDynamicShortcuts()

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    fun disableAppShortcut() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
+            shortcutManager.disableShortcuts(listOf(CONTACT_DIARY_SHORTCUT_ID))
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt:37]: As I understand our shortcuts are dynamic, they can be removed if not needed

        ShortcutManagerCompat.removeDynamicShortcuts()

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt [hunk 4]
```diff
@@ -89,6 +115,40 @@ class MainActivity : AppCompatActivity(), HasAndroidInjector {
...
+                )
+            )
+        } else {
+            navController.navigate(
+                ContactDiaryOnboardingFragmentDirections.actionContactDiaryOnboardingFragmentToContactDiaryOnboarding(
+                    showBottomNav = true,
+                    goToCurrentDay = true
+                )
+            )
+        }
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:151]: How do you feel about using navigating by Uri in this case with a custom schema such as corona-warn-app:// :
```kotlin
val request = NavDeepLinkRequest.Builder
    .fromUri("android-app://androidx.navigation.app/profile".toUri())
    .build()
findNavController().navigate(request)
```

File: Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_deltaonboarding.xml [hunk 5]
```diff
@@ -0,0 +1,153 @@
+<?xml version="1.0" encoding="utf-8"?>
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (d4rken) [inline Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_deltaonboarding.xml:2]: Can the `<layout` wrapper be removed?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt [hunk 6]
```diff
@@ -55,6 +76,11 @@ class MainActivity : AppCompatActivity(), HasAndroidInjector {
         factoryProducer = { viewModelFactory }
     )
 
+    private val homeFragmentVm: HomeFragmentViewModel by cwaViewModels(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:79]: Don't add the homefragment viewmodel here, MainActivity should not directly access that.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt [hunk 7]
```diff
@@ -108,6 +111,9 @@ class HomeFragment : Fragment(R.layout.home_fragment_layout), AutoInject {
                 HomeFragmentEvents.ShowReactivateRiskCheckDialog -> {
                     showReactivateRiskCheckDialog()
                 }
+                HomeFragmentEvents.GoToContactJournalDay -> {
+                    (activity as MainActivity).goToContactJournal()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:115]: Why does this logic have to go  through the home fragment, can't the activity directly navigate to the diary?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt (hunk 1 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt]: yes, good point I can use rather `ShortcutInfoCompat`

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt (hunk 2 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt (hunk 3 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt]: updated

File: Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_deltaonboarding.xml (hunk 5 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/deviceForTesters/res/layout/fragment_test_deltaonboarding.xml]: removed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt (hunk 7 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:115]: `HomeFragment` already handles `HomeFragmentEvents` such as delta onboarding or new release info and those events have priority before opening the contact journal. Maybe, I can even refactor this behavior and extract those events to `OnboardingActivity` and then I will be able to call the contact journal from `MainActivity`. Or let me know if you have another idea.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt (hunk 6 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:79]: I wanted to use the shared ViewModel for Activity > Fragment communication, the reason is explained in the following conversation

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt (hunk 4 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:151]: I don't know string URL could be error-prone, what would be the benefit of using the deep link within the app?

Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:151]: The benefit is that you can navigate deeply in the graph without having to pass some parameters from a screen into another in order to land on final screen. for example this flag ` goToCurrentDay = true` that you set in order to go the Day screen

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt (hunk 7 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:115]: This point also I think you can avoid using `navigate(uri())`

**Comment by mtwalli** [comment]
Two important thing to consider:
1- Restore the shortcuts if they are expected to be there, but actually they are not 
2- Create or (refresh) them in NON UI thread.

Please have a look into Google sample here :https://github.com/android/user-interface-samples/tree/main/AppShortcuts

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt (hunk 7 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt:115]: I've experimented with one concept and get cleaner results by refactoring onboarding, let me know your thoughts. It's in PR: https://github.com/corona-warn-app/cwa-app-android/pull/2398
Basically, I'm doing a check for delta onboarding and new features screen before starting `MainActivity` so it's guaranteed that those fragments will be shown before anything else

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt (hunk 4 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt:143]: `goToCurrentDay` parameter means that you should be redirect to `ContactDiaryDayFragment` after onboarding. We always want to show onboarding as a first page. I updated the code with deeplinks, let me know if you consider it better now.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt (hunk 6 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/MainActivity.kt]: updated

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt (hunk 7 above)
Comment (jurajkusnier) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragment.kt]: PR has been merged. So this comment is resolved now.

**Comment by jurajkusnier** [comment]
> Two important thing to consider:
> 1- Restore the shortcuts if they are expected to be there, but actually they are not
> 2- Create or (refresh) them in NON UI thread.
> 
> Please have a look into Google sample here :https://github.com/android/user-interface-samples/tree/main/AppShortcuts

I've took a look both points are covered in the code

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                .setIcon(IconCompat.createWithResource(context, R.drawable.ic_contact_diary_shortcut_icon))
+                .setIntent(createContactDiaryIntent())
+                .build()
+
+            ShortcutManagerCompat.addDynamicShortcuts(context, listOf(shortcut))
+        }
+    }
+
+    fun disableAppShortcut() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1) {
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/shortcuts/AppShortcutsHelper.kt:35]: This check is not needed, ShortcutManagerCompat.removeDynamicShortcuts already checks for it.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/resultavailable/SubmissionTestResultAvailableFragment.kt [hunk 9]
```diff
@@ -77,6 +79,7 @@ class SubmissionTestResultAvailableFragment : Fragment(R.layout.fragment_submiss
     override fun onResume() {
         super.onResume()
         binding.submissionTestResultAvailableContainer.sendAccessibilityEvent(AccessibilityEvent.TYPE_ANNOUNCEMENT)
+        appShortcutsHelper.disableAppShortcut()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/submission/resultavailable/SubmissionTestResultAvailableFragment.kt:82]: Please call it `removeAppShortcuts` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingLoadingFragment.kt [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+import de.rki.coronawarnapp.util.ui.doNavigate
+import de.rki.coronawarnapp.util.ui.observe2
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactoryProvider
+import de.rki.coronawarnapp.util.viewmodel.cwaViewModels
+import javax.inject.Inject
+
+class OnboardingLoadingFragment : Fragment(R.layout.onboaring_loading_layout), AutoInject {
+
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
+    private val vm: OnboardingLoadingViewModel by cwaViewModels(
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingLoadingFragment.kt:18]: Please rename to `viewModel` ?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingDeltaInteroperabilityFragment.kt [hunk 11]
```diff
@@ -53,7 +52,7 @@ class OnboardingDeltaInteroperabilityFragment :
 
         vm.navigateBack.observe2(this) {
             if (it) {
-                (requireActivity() as MainActivity).goBack()
+                activity?.onBackPressed()
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingDeltaInteroperabilityFragment.kt:55]: You can use `popBackStack()` ,I think it does the same

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingActivityModule.kt [hunk 12]
```diff
@@ -26,4 +28,12 @@ internal abstract class OnboardingActivityModule {
...
     @ContributesAndroidInjector(modules = [AnalyticsUIModule::class])
     abstract fun ppaUserInfoSelection(): AnalyticsUserInputFragment
+    @ContributesAndroidInjector(modules = [OnboardingLoadingModule::class])
+    abstract fun onboardingLoadingScreen(): OnboardingLoadingFragment
+    @ContributesAndroidInjector(modules = [NewReleaseInfoFragmentModule::class])
+    abstract fun newReleaseInfoFragment(): NewReleaseInfoFragment
+    @ContributesAndroidInjector(modules = [OnboardingDeltaInteroperabilityModule::class])
+    abstract fun onboardingDeltaInteroperabilityFragment(): OnboardingDeltaInteroperabilityFragment
+    @ContributesAndroidInjector(modules = [OnboardingDeltaAnalyticsModule::class])
+    abstract fun onboardingDeltaAnalyticsFragment(): OnboardingDeltaAnalyticsFragment
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/onboarding/OnboardingActivityModule.kt:38]: For readability can you please add Line breaks ?

File: Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml [hunk 13]
```diff
@@ -76,4 +91,46 @@
...
+            android:id="@+id/action_onboardingDeltaAnalyticsFragment_to_analyticsUserInputFragment"
+            app:destination="@id/analyticsUserInputFragment" />
+        <action
+            android:id="@+id/action_onboardingDeltaAnalyticsFragment_to_ppaMoreInfoFragment"
+            app:destination="@id/ppaMoreInfoFragment" />
+    </fragment>
+    <fragment
+        android:id="@+id/onboardingDeltaInteroperabilityFragment2"
+        android:name="de.rki.coronawarnapp.ui.onboarding.OnboardingDeltaInteroperabilityFragment"
+        android:label="OnboardingDeltaInteroperabilityFragment">
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml:125]: Please use `tools:layout` ?

File: Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml [hunk 14]
```diff
@@ -76,4 +91,46 @@
         android:name="de.rki.coronawarnapp.datadonation.analytics.ui.PpaMoreInfoFragment"
         android:label="PpaMoreInfoFragment"
         tools:layout="@layout/fragment_ppa_more_info" />
+
+    <!-- New Release -->
+    <fragment
+        android:id="@+id/newReleaseInfoFragment"
+        android:name="de.rki.coronawarnapp.release.NewReleaseInfoFragment"
+        android:label="NewReleaseInfoFragment">
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/nav_graph_onboarding.xml:99]: Please use `tools:layout`?

File: Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml [hunk 15]
```diff
@@ -87,6 +88,9 @@
             android:name="showBottomNav"
             android:defaultValue="true"
             app:argType="boolean" />
+        <deepLink app:uri="coronawarnapp://contact-journal/oboarding/?goToDay={goToDay}"
+            app:popUpTo="@id/contact_diary_nav_graph"
+            app:popUpToInclusive="true" />
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml:93]: Do you need it?

File: Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml [hunk 16]
```diff
@@ -19,6 +19,7 @@
         <action
             android:id="@+id/action_contactDiaryDayFragment_to_contactDiaryLocationBottomSheetDialogFragment"
             app:destination="@id/contactDiaryLocationBottomSheetDialogFragment" />
+        <deepLink app:uri="coronawarnapp://contact-journal/day/{selectedDay}" />
```
Comment (mtwalli) [inline Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml:22]: Do you need it too?

Comment (mtwalli) [reply Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml:22]: Sorry, I just saw it that you use it

File: Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml (hunk 15 above)
Comment (mtwalli) [reply Corona-Warn-App/src/main/res/navigation/contact_diary_nav_graph.xml:93]: Yes, I saw that are already using it

File: Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/deltaonboarding/ui/DeltaonboardingFragment.kt [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+import de.rki.coronawarnapp.util.ui.viewBindingLazy
+import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactoryProvider
+import de.rki.coronawarnapp.util.viewmodel.cwaViewModels
+import javax.inject.Inject
+
+@SuppressLint("SetTextI18n")
+class DeltaonboardingFragment : Fragment(R.layout.fragment_test_deltaonboarding), AutoInject {
+
+    @Inject lateinit var viewModelFactory: CWAViewModelFactoryProvider.Factory
+    private val vm: DeltaOnboardingFragmentViewModel by cwaViewModels { viewModelFactory }
```
Comment (mtwalli) [inline Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/deltaonboarding/ui/DeltaonboardingFragment.kt:20]: please rename to `viewModel` ?

**Comment by mtwalli** [comment]
Some small comment, otherwise looks really good 👍🏼.
Tested on Samsung Galaxy S8 Android 9 (Before and after the onboarding)

**Review by mtwalli — DISMISSED** [review]
Good Job!

**Comment by PhilippNowak96** [comment]
Tested on Pixel 4 XL Android 11, looks good! Thanks for integrating my initial commit as well! :)

**Review by d4rken — DISMISSED** [review]
lgtm

**Review by ralfgehrer — APPROVED** [review]
LGTM. Tested on Pixel (Android 10).

**Review by d4rken — APPROVED** [review]

**Review by mtwalli — APPROVED** [review]

**Review by janetback — APPROVED** [review]
UA reviewed.
